src/main/kotlin/com/atlassian/performance/tools/jvmtasks/api/TaskTimer.kt

```diff
@@ -17,12 +18,34 @@ object TaskTimer {
         task: () -> T
     ): T {
         CloseableThreadContext.push(label).use {
+            eventHandlers.forEach { it.onTaskStarted(label) }
             val start = now()
-            val result = task()
+            val result =
+                try {
+                    task()
+                } catch (e: Exception) {
+                    val duration = Duration.between(start, now())
+                    eventHandlers.forEach { it.onTaskFailed(label, duration) }
+                    throw e
+                }
             val duration = Duration.between(start, now())
-
+            eventHandlers.forEach { it.onTaskSucceeded(label, duration) }
             logger.debug("The task took $duration")
             return result
         }
     }
+
+    fun registerEventHandler(handler: EventHandler) {
+        eventHandlers.add(handler)
+    }
+
+    fun resetEventHandlers() {
+        eventHandlers.clear()
+    }
+
+    interface EventHandler {
```

Comment:
I like the "monohandlers" approach. It is [ISP](https://en.wikipedia.org/wiki/Interface_segregation_principle) friendly.

Reply:
> First event handlers in API. It's about time! ❤️

Personally I think it's a necessary evil. 

This `TaskTimer` is a global object, so we can't really inject watchable implementation of it to all its consumers. DI would be more managable approach. In such model we wouldn't need explicit event handlers/publishers, because you can just wrap the whole timer and see what starts and what ends. It's more powerful way of building software that allows to have any "event" type without explicit API for it.

Reply:
OTOH soon we'll have to emit custom events from all over the place, e.g. [SSH command messages](https://stash.atlassian.com/projects/JIRASERVER/repos/jira-performance-tests/pull-requests/1166/diff#gui/src/main/typescript/components/App.tsx?t=67) in [`facilitateSsh`](https://github.com/atlassian-labs/aws-resources/blob/c7f09f3166ffeb420e47101ac2018dc8fbddc7cd/src/main/kotlin/com/atlassian/performance/tools/aws/api/SshKeyFile.kt#L15). So we'll need a global pubsub event system for different kinds of publishers and subscribers.